baseplate/ratelimit.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import division
+
+import time
+
+from .context import ContextFactory
+
+
+class RateLimitExceededException(Exception):
+    """This exception gets raised whenever a rate limit is exceeded.
+    """
+    pass
+
+
+class RateLimiterContextFactory(ContextFactory):
+    """RateLimiter context factory
+
+    :param cache_context_factory: An instance of
+        :py:class:`baseplate.context.ContextFactory`.
+    :param ratelimit_cache_class: An instance of
+        :py:class:`baseplate.ratelimit.RateLimitCache`.
+    :param int allowance: The maximum allowance allowed per key.
+    :param int interval: The interval (in seconds) to reset allowances.
+    :param str key_prefix: A prefix to add to keys during rate limiting.
+        This is useful if you will have two different rate limiters that will
+        receive the same keys.
+
+    """
+
+    # TODO: This feels like a weird way of doing this
+    def __init__(self, cache_context_factory, ratelimit_cache_class,
+                 allowance=None, interval=None, key_prefix=''):
+        self.cache_context_factory = cache_context_factory
+        self.ratelimit_cache_class = ratelimit_cache_class
+        self.allowance = allowance
+        self.interval = interval
+        self.key_prefix = key_prefix
+
+    def make_object_for_context(self, name, server_span):
+        cache = self.cache_context_factory.make_object_for_context(
+            name, server_span)
+        ratelimit_cache = self.ratelimit_cache_class(cache)
+        return RateLimiter(ratelimit_cache, allowance=self.allowance,
+                           interval=self.interval, key_prefix=self.key_prefix)
+
+
+class RateLimiter(object):
+    """A class for rate limiting actions.
+
+    :param `RateLimitCache` cache: The backend to use for storing rate limit
+        counters.
+    :param int allowance: The maximum allowance allowed per key.
+    :param int interval: The interval (in seconds) to reset allowances.
+    :param str key_prefix: A prefix to add to keys during rate limiting.
+        This is useful if you will have two different rate limiters that will
+        receive the same keys.
+
+    """
+
+    def __init__(self, cache, allowance=None, interval=None,
```

Comment:
These shouldn't have default values if the default values are just going to raise exceptions.

Reply:
Hah. Yeah. Thanks for catching this